Modules/_sqlite/connection.c

```diff
@@ -168,9 +179,9 @@ pysqlite_connection_init_impl(pysqlite_Connection *self,
     self->thread_ident = PyThread_get_thread_ident();
     self->check_same_thread = check_same_thread;
 
-    self->function_pinboard_trace_callback = NULL;
-    self->function_pinboard_progress_handler = NULL;
-    self->function_pinboard_authorizer_cb = NULL;
+    SET_CALLBACK_CONTEXT(self->trace_ctx, NULL);
+    SET_CALLBACK_CONTEXT(self->progress_ctx, NULL);
+    SET_CALLBACK_CONTEXT(self->authorizer_ctx, NULL);
```

Comment:
These three statements are named `free_callback_contexts()` below. But here they are problematic.
* If this is a brand new connection, then `PyType_GenericAlloc` has set all these to NULL and they don't need to be NULLed again.
* If someone is calling `__init__` for the second time, this might free the Python callable, but SQLite will still have the pointer. Also the old `db` will leak, and who knows what'll happen to old cursors… Whoa, I don't see how reinitialization could work; should it rather raise an exception?

Reply:
I had reinitialisation in mind, but as you say, it's hard to imagine how to do that properly anyway. Maybe implicit connection close at the start of `__init__` will do the trick. I'll have to think that through. Perhaps we should fix the reinit case before proceeding with this PR.

Reply:
> Maybe implicit connection close at the start of `__init__` will do the trick.

I would be surprised. Cursors seem to have hard (and reasonable!) assumptions about their connection's `db`.
I also don't see the use case – when would a new connection object not be enough?

Reply:
Hm, true. Let's raise an exception instead.

Reply:
>  There's no refleaks on our side, and SQLite seems to be able to clean up as well. 

I'm surprised. Wouldn't this result in a `sqlite3_open_v2` without a corresponding close?

Reply:
Perhaps I was a little bit vague. I did the tests with an added implicit close in `__init__`.

Reply:
I see. Well, this is more complicated than I initially thought; I guess it should be a separate bpo issue.

I keep finding issues:

- `db` is not set to NULL if init fails.
- This segfaults for me:

```python
import sqlite3

conn = sqlite3.connect(':memory:')
conn.execute('CREATE TABLE foo (bar)')

try:
    conn.__init__('/bad-file/')
except sqlite3.OperationalError:
    pass

conn.execute('INSERT INTO foo (bar) VALUES (1), (2), (3), (4)')
```
... and even if that's fixed, I'm concerned that there will be other edge cases we can't reasonably think about.

Reply:
Yes, there's some issues in connection init. I'd gladly clean it up and harden it before continuing with this PR.

Reply:
It's not a blocker for this PR; just please open a bpo so the issues won't be lost in a closed PR.

Reply:
I've opened issue 45126 for this.